### PR TITLE
Improve plugins grouping (alternative to #1424)

### PR DESCRIPTION
**Pull request by buixor**
Fixes https://discourse.crowdsec.net/t/http-plugin-configuration-group-wait-and-group-threshold-problem/649/3

This error was caused due to somewhat wrong default config.

New logic is:
 - Simplify the watcher : only the ticker-based watcher can trigger events processing
 - Ensure w/ extra tests that treshold can be set by "time", "count", "time+count" or none

### CODE REVIEW

File: pkg/csplugin/broker_test.go [hunk 1]
```diff
@@ -252,7 +255,284 @@ func TestBrokerInit(t *testing.T) {
...
+	log.Printf("second send")
+	pb.PluginChannel <- ProfileAlert{ProfileID: uint(0), Alert: &models.Alert{}}
+	log.Printf("third send")
+	pb.PluginChannel <- ProfileAlert{ProfileID: uint(0), Alert: &models.Alert{}}
+	time.Sleep(500 * time.Millisecond)
+	//because of group threshold, we shouldn't have data yet
+	assert.NoFileExists(t, "./out")
+	log.Printf("before sleep")
+	time.Sleep(1 * time.Second)
+	//after 8 seconds, we should have data
```
Comment (sbs2001) [inline pkg/csplugin/broker_test.go:371]: Comment doesn't seem to relevant anymore

File: pkg/csplugin/broker_test.go [hunk 2]
```diff
@@ -252,7 +255,284 @@ func TestBrokerInit(t *testing.T) {
...
+
+func TestBrokerNoThreshold(t *testing.T) {
+	var alerts []models.Alert
+	DefaultEmptyTicker = 50 * time.Millisecond
+
+	buildDummyPlugin()
+	setPluginPermTo744()
+	defer tearDown()
+	//init
+	procCfg := csconfig.PluginCfg{}
```
Comment (sbs2001) [inline pkg/csplugin/broker_test.go:288]: :nit Name and value don't match.

File: pkg/csplugin/broker_test.go [hunk 3]
```diff
@@ -252,7 +255,284 @@ func TestBrokerInit(t *testing.T) {
...
+
+func TestBrokerRunGroupAndTimeThreshold_TimeFirst(t *testing.T) {
+	//test grouping by "time"
+	DefaultEmptyTicker = 50 * time.Millisecond
+	buildDummyPlugin()
+	setPluginPermTo744()
+	defer tearDown()
+
+	//init
+	procCfg := csconfig.PluginCfg{}
```
Comment (sbs2001) [inline pkg/csplugin/broker_test.go:340]: :nit Name and value don't match.

File: pkg/csplugin/broker_test.go [hunk 4]
```diff
@@ -252,7 +255,284 @@ func TestBrokerInit(t *testing.T) {
...
+	}
+}
+
+func TestBrokerRunGroupAndTimeThreshold_CountFirst(t *testing.T) {
+	DefaultEmptyTicker = 50 * time.Millisecond
+	buildDummyPlugin()
+	setPluginPermTo744()
+	defer tearDown()
+	//init
+	procCfg := csconfig.PluginCfg{}
```
Comment (sbs2001) [inline pkg/csplugin/broker_test.go:392]: :nit Name and value don't match.

File: pkg/csplugin/broker_test.go [hunk 5]
```diff
@@ -252,7 +255,284 @@ func TestBrokerInit(t *testing.T) {
...
+
+func TestBrokerRunGroupThreshold(t *testing.T) {
+	//test grouping by "size"
+	DefaultEmptyTicker = 50 * time.Millisecond
+	buildDummyPlugin()
+	setPluginPermTo744()
+	defer tearDown()
+
+	//init
+	procCfg := csconfig.PluginCfg{}
```
Comment (sbs2001) [inline pkg/csplugin/broker_test.go:447]: :nit Name and value don't match.

File: pkg/csplugin/broker_test.go [hunk 6]
```diff
@@ -252,7 +255,284 @@ func TestBrokerInit(t *testing.T) {
...
+	}
+}
+
+func TestBrokerRunTimeThreshold(t *testing.T) {
+	DefaultEmptyTicker = 50 * time.Millisecond
+	buildDummyPlugin()
+	setPluginPermTo744()
+	defer tearDown()
+	//init
+	procCfg := csconfig.PluginCfg{}
```
Comment (sbs2001) [inline pkg/csplugin/broker_test.go:495]: :nit Name and value don't match.

**Review by sbs2001 — CHANGES_REQUESTED** [review]
Some nitpicks about style. Tests cases look good to me

File: pkg/csplugin/broker_test.go [hunk 7]
```diff
@@ -276,10 +556,16 @@ func TestBrokerRun(t *testing.T) {
...
-	assert.FileExists(t, "./out")
-	assert.Equal(t, types.GetLineCountForFile("./out"), 2)
+	content, err := ioutil.ReadFile("./out")
+	if err != nil {
+		log.Errorf("Error reading file: %s", err)
+	}
+	var alerts []models.Alert
+	err = json.Unmarshal(content, &alerts)
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(alerts))
```
Comment (sbs2001) [inline pkg/csplugin/broker_test.go:568]: Just thinking out loud. This means Threshold=1 is not honored and the watcher went with ticker event. Had it followed threshold = 1 then plugin would've fired 2 separate events and `assert.Equal(t, types.GetLineCountForFile("./out"), 2)` would pass.

File: pkg/csplugin/watcher.go [hunk 8]
```diff
@@ -42,15 +50,62 @@ func (pw *PluginWatcher) Start(tomb *tomb.Tomb) {
...
+		//only time is set
+		watchTime = interval
+	} else if interval != 0 && threshold != 0 {
+		//both are set
+		watchTime = DefaultEmptyTicker
+		watchCount = threshold
+	} else {
+		//none are set, we sent every event we receive
+		watchTime = DefaultEmptyTicker
+		watchCount = 1
```
[the remark this answers is not in the document]
Comment (buixor) [reply pkg/csplugin/watcher.go:76]: which one you think is best?

File: pkg/csplugin/broker_test.go (hunk 7 above)
Comment (buixor) [reply pkg/csplugin/broker_test.go:556]: It's because both arrive in burst, we already have another test checking that one item .. sleep ...  one item leads to two different writes

**Review by sabban — APPROVED** [review]
lgtm

**Review by sbs2001 — APPROVED** [review]
